sycl/test-e2e/Basic/built-ins/helpers.hpp

```diff
@@ -30,29 +30,30 @@ void test(bool CheckDevice, double delta, FuncTy F, ExpectedTy Expected,
 
   if (!CheckDevice)
     return;
+  bool result;
+  {
+    sycl::buffer<bool, 1> SuccessBuf{&result, sycl::range<1>{1}};
 
-  sycl::buffer<bool, 1> SuccessBuf{1};
-
-  // Make sure we don't use fp64 on devices that don't support it.
-  sycl::detail::get_elem_type_t<ExpectedTy> d(delta);
-
-  sycl::queue{}.submit([&](sycl::handler &cgh) {
```

Comment:
I wonder if we actually have a bug somewhere in our execution graph builder. [`queue` destruction](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interface.queue.class) is a non-blocking operation, but the kernel should still be launched and all completion events communicated as usual:

> A SYCL queue may be destroyed even when there are uncompleted [commands](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#command) that have been submitted to the queue. Doing so does not block. Instead, any commands that have been submitted to the queue begin execution when their requisites are satisfied, just as they would had the queue not been destroyed. Any event objects for those commands are signaled in the normal manner when the command completes. Resources associated with the queue will be freed by the time the last command completes.

Which makes me think that it could be a bug that we don't communicate kernel completion event properly and `host_accessor` creation doesn't wait for it

Reply:
I don't think this temporary queue is the problem here to be honest, I just rewrote it to declare it beforehand in order to improve readability. The change that actually fixed the test was introducing the new boolean variable `result` and having the buffer point to it. This is also unusual because according to the spec, host accessor can be safely used to access buffers like the test was doing before and yet it was failing so we could also have a bug in host accessor.

Reply:
> host accessor can be safely used to access buffers like the test was doing before and yet it was failing so we could also have a bug in host accessor

IMO, we shouldn't be merging this "fix" until the investigation is done.

Reply:
I did some more investigation, it seems to be a GPU accuracy issue. Note in line 37 of helpers.hpp the delta error tolerance is converted from double to whatever the type that the function under test produces. If this type happens to be float, then some accuracy is lost going from double to float and apparently in some of these cases, the results of the GPU computation differ from the expected values by large enough errors so as to expose this loss in accuracy and the equal(result, expected, delta) function that verifies the result return false which causes our assertion to fail.

Therefore, the synchronization is actually correct, the problem seems to be the lack of accuracy of GPU for float arguments. Easy fix by removing the variable line 37 and replacing d by the original delta which has type double?

Reply:
I agree with the idea of upscaling before error-tolerance checking.

Reply:
I've made the changes. Also, I explicitly created a context and created a queue with that context in order to make the test 
 independent on the default context extension.